### PR TITLE
[Graphbolt] Add negative sampler function in csc graph

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add uniform negative sample function in csc sampling graph.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 1]
```diff
@@ -150,6 +150,28 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
...
+   * in the graph.
+   * @param negative_ratio The ratio of the number of negative samples to
+   * positive samples.
+   *
+   * @return A tuple consisting of two 1D tensors represents the source and
+   * destination of negative edges. Note that negative refers to false
+   * negatives, which means the edge could be present or not present in the
+   * graph.
+   */
+  std::tuple<torch::Tensor, torch::Tensor> SampleNegativePerSourceUniform(
```
Comment (frozenbugs) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:171]: what about std::pair?

Comment (frozenbugs) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:171]: What about SampleNegativeEdgesUniform? Do we have other negative sampling method other than per source?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 2]
```diff
@@ -296,6 +296,38 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_negative_per_source_uniform(self, pos_pairs, negative_ratio):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:299]: ditto.

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 1 above)
Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:171]: I'm not sure `std::pair` can be identified in the python side, as this function is an pybind function.

Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:171]: There could be a function like `SampleNegativePerDestinationUniform`, but we don't need it at present. So it's fine to change the name to `SampleNegativeEdgesUniform`.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 2 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:299]: fixed.

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 1 above)
Comment (frozenbugs) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:171]: I see, thanks for the clarification.

Comment (frozenbugs) [reply graphbolt/include/graphbolt/csc_sampling_graph.h]: I see, per source is more common AFAIK, let's keep it like this and revisit in the future.

**Review by frozenbugs — APPROVED** [review]

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 3]
```diff
@@ -162,6 +162,8 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
    * in the graph.
    * @param negative_ratio The ratio of the number of negative samples to
    * positive samples.
+   * @param negative_ratio The total count of nodes that can be selected.
```
Comment (frozenbugs) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:165]:
```suggestion
   * @param num_nodes The total count of nodes that can be selected.
```

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 4]
```diff
@@ -300,7 +300,9 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
-    def sample_negative_edges_uniform(self, pos_pairs, negative_ratio):
+    def sample_negative_edges_uniform(
+        self, pos_pairs, negative_ratio, edge_type
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:304]:
```suggestion
        self, node_pairs, negative_ratio, edge_type
```

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 3 above)
Comment (frozenbugs) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:165]: total count of dst nodes per src node?

Comment (frozenbugs) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:165]: We need better name than num_nodes.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:304]: fixed

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 3 above)
Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:165]: In general it use `negative_ratio` in GNN. And in the future we may support `float` type, so a ratio is more fit?

Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:165]: This does have some ambiguity. Actually it means the max node id can be selected from, ususally it is the number of a specific type node. So use `max_node_id  ` is better?

File: graphbolt/src/csc_sampling_graph.cc [hunk 5]
```diff
@@ -195,6 +195,18 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighbors(
       subgraph_reverse_edge_ids, subgraph_type_per_edge);
 }
 
+std::tuple<torch::Tensor, torch::Tensor>
+CSCSamplingGraph::SampleNegativeEdgesUniform(
+    const std::tuple<torch::Tensor, torch::Tensor>& pos_pairs,
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/src/csc_sampling_graph.cc:200]: fixed

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 6]
```diff
@@ -150,6 +150,30 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
...
+   * @param max_node_id The maximum ID of the node to be selected. It
+   * should correspond to the number of nodes of a specific type.
+   *
+   * @return A tuple consisting of two 1D tensors represents the source and
+   * destination of negative edges. Note that negative refers to false
+   * negatives, which means the edge could be present or not present in the
+   * graph.
+   */
+  std::tuple<torch::Tensor, torch::Tensor> SampleNegativeEdgesUniform(
+      const std::tuple<torch::Tensor, torch::Tensor>& pos_pairs,
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:174]: fixed

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 7]
```diff
@@ -300,6 +300,51 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_negative_edges_uniform(
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:303]: It is in later PR #5923

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 8]
```diff
@@ -300,6 +300,51 @@ def sample_neighbors(
...
+        Tuple[Tensor]
+            A tuple consisting of two 1D tensors represents the source and
+            destination of negative edges. Note that negative refers to false
+            negatives, which means the edge could be present or not present in
+            the graph.
+        """
+        if edge_type:
+            _, _, dst_node_type = edge_type
+            dst_node_type_id = self.metadata.node_type_to_id[dst_node_type]
+            num_nodes = (
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:336]: When `edge_type` specified, it should be the number of nodes of the edge type, `node_type_offset` is required in this case.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 9]
```diff
@@ -304,6 +304,59 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_negative_edges_uniform(
+        self, node_pairs, negative_ratio, edge_type
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:308]: Let's order the parameters in a more natural way. edge_type decides how to understand node_pairs, it should group with node_pairs, and negative_ratio indicates the operator details.
```suggestion
        self, edge_type, node_pairs, negative_ratio
```

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 10]
```diff
@@ -304,6 +304,59 @@ def sample_neighbors(
...
+            sampled will also have the same type. If set to None, it will be
+            considered as a homogeneous graph.
+
+        Returns
+        -------
+        Tuple[Tensor]
+            A tuple consisting of two 1D tensors represents the source and
+            destination of negative edges. In the context of a heterogeneous
+            graph, both the input nodes and the selected nodes are represented
+            by heterogeneous IDs. Note that negative refers to false negatives,
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:338]: Worth to mention that the edge type of the tuple will be the input edge type.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 9 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:308]: Make sense, changed for fix

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 10 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py]: added

**Comment by frozenbugs** [comment]
@dgl-bot

**Pull request merged**